### PR TITLE
Survivorship proof-of-concept implementation and optional manual override for GR merge

### PR DESCRIPTION
**Pull request by nvg-smile**
MDM requires a mechanism to determine which field values will be updated in the golden record. IMdmSurvivorshipService provides a way to configure that. Default implementation uses EMPI merge rules - overwriting GR values on updates and combining them on merge. Additionally, GR values can be manually overridden on merge.

### CODE REVIEW

File: hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/svc/MdmGoldenResourceMergerSvcTest.java [hunk 1]
```diff
@@ -376,39 +395,37 @@ public void from12To23() {
...
+
+		myToGoldenPatient.addName().addGiven("Jeff");
+		myToGoldenPatient.getNameFirstRep().addGiven("George");
+		assertThat(myToGoldenPatient.getName(), hasSize(1));
+		assertThat(myToGoldenPatient.getName().get(0).getGiven(), hasSize(2));
+
+		Patient mergedSourcePatient = mergeGoldenPatients();
+		assertThat(mergedSourcePatient.getName(), hasSize(2));
+		assertThat(mergedSourcePatient.getName().get(0).getGiven(), hasSize(2));
+		assertThat(mergedSourcePatient.getName().get(1).getGiven(), hasSize(2));
```
Comment (tadgh) [inline hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/svc/MdmGoldenResourceMergerSvcTest.java:411]: We should also assert on the actual contents of these givens

File: hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/svc/MdmSurvivorshipSvcImplTest.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package ca.uhn.fhir.jpa.mdm.svc;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class MdmSurvivorshipSvcImplTest {
+
+}
```
Comment (tadgh) [inline hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/svc/MdmSurvivorshipSvcImplTest.java:7]: empty class? or did you mean to add tests here?

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/util/TerserUtil.java [hunk 3]
```diff
@@ -84,23 +124,143 @@ public static void cloneCompositeField(FhirContext theFhirContext, IBaseResource
...
 
-	private static boolean contains(IBase theItem, List<IBase> theItems) {
+	private static boolean containsPrimitiveValue(IBase theItem, List<IBase> theItems) {
 		PrimitiveTypeEqualsPredicate predicate = new PrimitiveTypeEqualsPredicate();
 		return theItems.stream().anyMatch(i -> {
 			return predicate.test(i, theItem);
 		});
 	}
 
+	private static boolean contains(IBase theItem, List<IBase> theItems) {
```
Comment (tadgh) [inline hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/util/TerserUtil.java:145]: You should probably have james or ken review the terser chunk here

**Review by tadgh — APPROVED** [review]
Approved pending todos, and a subsequent review of the terser stuff from james/ken

File: hapi-fhir-jpaserver-mdm/src/main/java/ca/uhn/fhir/jpa/mdm/svc/MdmLinkUpdaterSvcImpl.java [hunk 4]
```diff
@@ -89,6 +93,13 @@ public IAnyResource updateLink(IAnyResource theGoldenResource, IAnyResource theS
 		mdmLink.setMatchResult(theMatchResult);
 		mdmLink.setLinkSource(MdmLinkSourceEnum.MANUAL);
 		myMdmLinkDaoSvc.save(mdmLink);
+
+		// IAnyResource resource = (theManuallyMergedGoldenResource == null) ? theSourceResource : theManuallyMergedGoldenResource;
```
[the remark this answers is not in the document]
Comment (nvg-smile) [reply hapi-fhir-jpaserver-mdm/src/main/java/ca/uhn/fhir/jpa/mdm/svc/MdmLinkUpdaterSvcImpl.java:97]: Fixed

File: hapi-fhir-jpaserver-mdm/src/main/java/ca/uhn/fhir/jpa/mdm/svc/MdmSurvivorshipSvcImpl.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import org.hl7.fhir.instance.model.api.IBaseResource;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class MdmSurvivorshipSvcImpl implements IMdmSurvivorshipService {
+
+	@Autowired
+	private FhirContext myFhirContext;
+
+	/**
+	 * TODO NG Update docs to tell taht we also do that in non GR merge - like linking
```
[the remark this answers is not in the document]
Comment (nvg-smile) [reply hapi-fhir-jpaserver-mdm/src/main/java/ca/uhn/fhir/jpa/mdm/svc/MdmSurvivorshipSvcImpl.java:37]: Updated

File: hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/svc/MdmGoldenResourceMergerSvcTest.java (hunk 1 above)
Comment (nvg-smile) [reply hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/svc/MdmGoldenResourceMergerSvcTest.java:411]: Fixed

File: hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/svc/MdmSurvivorshipSvcImplTest.java (hunk 2 above)
Comment (nvg-smile) [reply hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/svc/MdmSurvivorshipSvcImplTest.java:7]: Fixed

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/util/TerserUtil.java (hunk 3 above)
Comment (nvg-smile) [reply hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/util/TerserUtil.java:149]: discussed in a meeting - added more tests to cover more involved use cases

**Pull request merged**